model/project_parser_db.go

```diff
@@ -61,56 +65,101 @@ func ParserProjectUpdateOne(query interface{}, update interface{}) error {
+// existing config number is less than or equal to the new config number.
+// If shouldEqual, only update if the config update number matches.
+func (pp *ParserProject) UpsertWithConfigNumber(num int, shouldEqual bool) error {
 	if pp.Id == "" {
 		return errors.New("no version ID given")
 	}
-
+	pp.ConfigUpdateNumber = num + 1
 	found, err := ParserProjectFindOneById(pp.Id)
 	if err != nil {
 		return errors.Wrapf(err, "error finding parser project '%s'", pp.Id)
 	}
+	// the document doesn't exist yet, so just insert it
 	if found == nil {
-		pp.ConfigUpdateNumber = num
-		return errors.Wrap(pp.Insert(), "error inserting parser project")
+		if err = pp.Insert(); err == nil {
+			return nil
+		}
+		grip.Debug(message.WrapError(err, message.Fields{
+			"message":                         "new parser project could not be inserted",
+			"version":                         pp.Id,
+			"attempted_to_update_with_number": pp.ConfigUpdateNumber,
+		}))
+		// if the insert didn't succeed, likely this ID has now already been inserted, so we should overwrite it
 	}
 
-	return errors.Wrapf(ParserProjectUpdateOne(
-		bson.M{
-			ParserProjectIdKey: pp.Id,
-			"$or": []bson.M{
-				bson.M{ParserProjectConfigNumberKey: bson.M{"$exists": false}},
-				bson.M{ParserProjectConfigNumberKey: pp.ConfigUpdateNumber},
-			},
-		},
-		bson.M{
-			"$set": bson.M{
-				ParserProjectEnabledKey:           pp.Enabled,
-				ParserProjectStepbackKey:          pp.Stepback,
-				ParserProjectPreErrorFailsTaskKey: pp.PreErrorFailsTask,
-				ParserProjectBatchTimeKey:         pp.BatchTime,
-				ParserProjectOwnerKey:             pp.Owner,
-				ParserProjectRepoKey:              pp.Repo,
-				ParserProjectRepoKindKey:          pp.RepoKind,
-				ParserProjectRemotePathKey:        pp.RemotePath,
-				ParserProjectBranchKey:            pp.Branch,
-				ParserProjectIdentifierKey:        pp.Identifier,
-				ParserProjectDisplayNameKey:       pp.DisplayName,
-				ParserProjectCommandTypeKey:       pp.CommandType,
-				ParserProjectIgnoreKey:            pp.Ignore,
-				ParserProjectPreKey:               pp.Pre,
-				ParserProjectPostKey:              pp.Post,
-				ParserProjectTimeoutKey:           pp.Timeout,
-				ParserProjectCallbackTimeoutKey:   pp.CallbackTimeout,
-				ParserProjectModulesKey:           pp.Modules,
-				ParserProjectBuildVariantsKey:     pp.BuildVariants,
-				ParserProjectFunctionsKey:         pp.Functions,
-				ParserProjectTaskGroupsKey:        pp.TaskGroups,
-				ParserProjectTasksKey:             pp.Tasks,
-				ParserProjectExecTimeoutSecsKey:   pp.ExecTimeoutSecs,
-				ParserProjectLoggersKey:           pp.Loggers,
-				ParserProjectConfigNumberKey:      num,
-			},
-		}), "error updating parser project '%s'", pp.Id)
+	q := bson.M{ParserProjectIdKey: pp.Id}
+	if shouldEqual { // guarantee that the config number hasn't changed
+		q["$or"] = []bson.M{
+			bson.M{ParserProjectConfigNumberKey: bson.M{"$exists": false}},
```

Comment:
Shouldn't this only be in the case where the config number key is 0? If it's some number, we shouldn't check `"$exists": false`, since that implies 0.